sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala

```diff
@@ -218,24 +218,20 @@ private[sql] case class InsertIntoHadoopFsRelation(
         val partitionProj = newProjection(codegenEnabled, partitionOutput, output)
         val dataProj = newProjection(codegenEnabled, dataOutput, output)
 
-        if (needsConversion) {
-          val converter = CatalystTypeConverters.createToScalaConverter(dataSchema)
-          while (iterator.hasNext) {
-            val row = iterator.next()
-            val partitionPart = partitionProj(row)
-            val dataPart = dataProj(row)
-            val convertedDataPart = converter(dataPart).asInstanceOf[Row]
-            writerContainer.outputWriterForRow(partitionPart).write(convertedDataPart)
-          }
+        val partitionSchema = StructType.fromAttributes(partitionOutput)
+        val partConverter = CatalystTypeConverters.createToScalaConverter(partitionSchema)
```

Comment:
Actually, is there ever a case that we actually need to convert the result of the partitioning projection?  The only case that I can think of would be a situation where the projection is a no-op that returns the original row and we apply it to a Scala row that hasn't undergone conversion.  If that's the case, though, why did the old code only convert the projection in the `!needsConversion` branch?